### PR TITLE
[WIP] Onedir build

### PR DESCRIPTION
**Pull request by dwoz**
### What does this PR do?

This is an example onedir build using make to compile python and install salt. The build process does the following to make things portable:

- Modify rpath of shared libraries to look in `$ORIGIN` for other shared libraries being loaded.
- Custom shebang for installed scripts to make sure the proper python binary is used
- Customize sys.prefix and sys.exe_prefix to make sure proper python location is used
- Pin sys.path into the proper python location

### What issues does this PR fix or reference?
Fixes:

### Previous Behavior
Remove this section if not relevant

### New Behavior
Remove this section if not relevant

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

Needs docs, tests changelog

- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -958,6 +982,12 @@ def salt_master_factory(
     config_defaults["known_hosts_file"] = tests_known_hosts_file
     config_defaults["syndic_master"] = "localhost"
     config_defaults["transport"] = salt_syndic_master_factory.config["transport"]
+    config_defaults["engines_dirs"] = [
+        str(salt_factories.get_salt_engines_path()),
+    ]
+    config_defaults["log_handlers_dirs"] = [
+        str(salt_factories.get_salt_log_handlers_path()),
+    ]
```
Comment (s0undt3ch) [inline tests/conftest.py:990]:
```suggestion
```
We should no longer require this.

File: tests/conftest.py [hunk 2]
```diff
@@ -676,6 +689,17 @@ def groups_collection_modifyitems(config, items):
...
+
+# Hook the system installation of salt into salt_factories
+@pytest.fixture(scope="session")
+def salt_factories(salt_factories, request):
+    scripts_path = request.config.getoption("--scripts-path")
+    if scripts_path:
+        salt_factories.scripts_dir = pathlib.Path(scripts_path)
+    return salt_factories
+
+
```
Comment (s0undt3ch) [inline tests/conftest.py:702]:
```suggestion
```
And instead:

```python
@pytest.fixture(scope="session")
def salt_factories_config():
    """
    Return a dictionary with the keyworkd arguments for FactoriesManager
    """
    return {
        "code_dir": str(CODE_DIR),
        "inject_coverage": MAYBE_RUN_COVERAGE,
        "inject_sitecustomize": MAYBE_RUN_COVERAGE,
        "scripts_path": request.config.getoption("--scripts-path"),
        "start_timeout": 120
        if (os.environ.get("JENKINS_URL") or os.environ.get("CI"))
        else 60,
    }
```

File: tests/pytests/conftest.py [hunk 3]
```diff
@@ -149,6 +149,12 @@ def salt_master_factory(
     config_defaults["reactor"] = [
         {reactor_event.tag: [str(reactor_event.sls_path)]},
     ]
+    config_defaults["engines_dirs"] = [
+        str(salt_factories.get_salt_engines_path()),
+    ]
+    config_defaults["log_handlers_dirs"] = [
+        str(salt_factories.get_salt_log_handlers_path()),
+    ]
```
Comment (s0undt3ch) [inline tests/pytests/conftest.py:157]:
```suggestion
```

File: tests/pytests/conftest.py [hunk 4]
```diff
@@ -308,6 +316,12 @@ def salt_minion_factory(salt_master_factory, salt_minion_id, sdb_etcd_port, vaul
         "auth": {"method": "token", "token": "testsecret", "uses": 0},
         "policies": ["testpolicy"],
     }
+    config_defaults["engines_dirs"] = [
+        str(salt_factories.get_salt_engines_path()),
+    ]
+    config_defaults["log_handlers_dirs"] = [
+        str(salt_factories.get_salt_log_handlers_path()),
+    ]
```
Comment (s0undt3ch) [inline tests/pytests/conftest.py:324]:
```suggestion
```

File: tests/pytests/conftest.py [hunk 5]
```diff
@@ -330,14 +344,20 @@ def salt_minion_factory(salt_master_factory, salt_minion_id, sdb_etcd_port, vaul
...
         config_defaults = yaml.deserialize(rfh.read())
     config_defaults["hosts.file"] = os.path.join(RUNTIME_VARS.TMP, "hosts")
     config_defaults["aliases.file"] = os.path.join(RUNTIME_VARS.TMP, "aliases")
     config_defaults["transport"] = salt_master_factory.config["transport"]
+    config_defaults["engines_dirs"] = [
+        str(salt_factories.get_salt_engines_path()),
+    ]
+    config_defaults["log_handlers_dirs"] = [
+        str(salt_factories.get_salt_log_handlers_path()),
+    ]
```
Comment (s0undt3ch) [inline tests/pytests/conftest.py:360]:
```suggestion
```

File: pkg/onedir.sh [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+# onedir.sh
+#
+# Script for making a onedir build. Requires Docker
+#
+#
+set +e
+
+# XXX: centos 7 with dependencies installed, we need to figure out how we'll
+# handle this
+[ -z "$CONTAINER_IMAGE" ] && CONTAINER_IMAGE=dwoz1/cicd:onedir-centos7
```
Comment (ITJamie) [inline pkg/onedir.sh:12]: Should these containers not exist in a saltstack docker hub rather than a personal docker hub account?

Comment (dwoz) [reply pkg/onedir.sh:12]: Yes, any docker images that are being used here will eventually be moved under the saltstack org. 

The repo for the current containers is here: https://github.com/dwoz/docker-dwoz1

**Pull request closed without merging**